File: src/main/java/leetcode/topinterview/LeetCode4.java
package leetcode.topinterview;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/21 01:40
 */
public class LeetCode4 {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        final int n1 = nums1.length;
        final int n2 = nums2.length;
        if (n1 > n2) {
            return findMedianSortedArrays(nums2, nums1);
        }
        int l = 0;
        int r = n1;
        while (l <= r) {
            int cutA = (l + r) / 2;
            int cutB = (n1 + n2 + 1) / 2 - cutA;
            final int L1 = cutA == 0 ? Integer.MIN_VALUE : nums1[cutA - 1];
            final int L2 = cutB == 0 ? Integer.MIN_VALUE : nums2[cutB - 1];
            final int R1 = cutA == n1 ? Integer.MAX_VALUE : nums1[cutA];
            final int R2 = cutB == n2 ? Integer.MAX_VALUE : nums2[cutB];
            if (L1 > R2) {
                r = cutA - 1;
            } else if (L2 > R1) {
                l = cutA + 1;
            } else {
                return (n1 + n2) % 2 == 0 ? (Math.max(L1, L2) + Math.min(R1, R2)) * 0.5 : Math.max(L1, L2);
            }
        }
        return -1;
    }
}

File: src/main/go/leetcode/lists/hot100/LeetCode4.go
package hot100

import "math"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	n1, n2 := len(nums1), len(nums2)
	if n1 > n2 {
		return findMedianSortedArrays(nums2, nums1)
	}
	length := n1 + n2
	l, r := 0, n1
	for l <= r {
		cutA := (l + r) / 2
		cutB := (length+1)/2 - cutA //左半边比右半边多一个元素
		var (
			L1 int
			R1 int
			L2 int
			R2 int
		)
		if cutA == 0 {
			L1 = math.MinInt32
		} else {
			L1 = nums1[cutA-1]
		}
		if cutB == 0 {
			L2 = math.MinInt32
		} else {
			L2 = nums2[cutB-1]
		}
		if cutA == n1 {
			R1 = math.MaxInt32
		} else {
			R1 = nums1[cutA]
		}
		if cutB == n2 {
			R2 = math.MaxInt32
		} else {
			R2 = nums2[cutB]
		}
		if L2 <= R1 && L1 <= R2 {
			if (n1+n2)%2 == 0 {
				return (float64(Max(L1, L2)) + float64(Min(R1, R2))) * 0.5
			}
			return float64(Max(L1, L2))
		} else if L1 > R2 {
			r = cutA - 1
		} else if L2 > R1 {
			l = cutA + 1
		}
	}
	return -1
}
